config/config.go

```diff
@@ -166,14 +260,384 @@ func (cs configs) ResolveConfig(f func(configs []interface{}) error) error {
 // The default configuration sources are:
 // * Environment Variables
 // * Shared Configuration and Shared Credentials files.
-func LoadDefaultConfig(cfgs ...Config) (aws.Config, error) {
-	var cfgCpy configs
-	cfgCpy = append(cfgCpy, cfgs...)
+func LoadDefaultConfig(ctx context.Context, optFns ...func(*LoadOptions) error) (cfg aws.Config, err error) {
+	var options LoadOptions
+	for _, optFn := range optFns {
+		optFn(&options)
+	}
+
+	// assign Load Options to configs
+	var cfgCpy = configs{options}
+
+	cfgCpy, err = cfgCpy.AppendFromLoaders(ctx, defaultLoaders)
+	if err != nil {
+		return aws.Config{}, err
+	}
 
-	cfgCpy, err := cfgCpy.AppendFromLoaders(defaultLoaders)
+	cfg, err = cfgCpy.ResolveAWSConfig(ctx, defaultAWSConfigResolvers)
 	if err != nil {
 		return aws.Config{}, err
 	}
 
-	return cfgCpy.ResolveAWSConfig(defaultAWSConfigResolvers)
+	return cfg, nil
+}
+
+// getRegion returns Region from config's LoadOptions
+func (o LoadOptions) getRegion(ctx context.Context) (string, bool, error) {
+	if len(o.Region) == 0 {
+		return "", false, nil
+	}
+
+	return o.Region, true, nil
+}
+
+// WithRegion is a helper function to construct functional options
+// that sets Region on config's LoadOptions. Setting the region to
+// an empty string, will result in the region value being ignored.
+func WithRegion(v string) LoadOptionsFunc {
+	return func(o *LoadOptions) error {
+		o.Region = v
+		return nil
+	}
+}
+
+// getDefaultRegion returns DefaultRegion from config's LoadOptions
+func (o LoadOptions) getDefaultRegion(ctx context.Context) (string, bool, error) {
+	if len(o.DefaultRegion) == 0 {
+		return "", false, nil
+	}
+
+	return o.DefaultRegion, true, nil
+}
```

Comment:
Should `getDefaultRegion` be used to resolve the EC2 IMDS Region, if no other region is found, `UseEC2IMDSRegion` option if its not nil?

**OR** Is there somewhere else in the "chain" of configuration sources where `UseEC2IMDSRegion` option should be used?

Probably will want to document cleary on that member when in the order of configuration sources the SDK will decide to get the region from the EC2 instance.


I might be missing it but i'm not seeing how the `UseEC2IMDSRegion` option is used to derive a region.

Reply:
UseEC2IMDSRegion isn't resolved at the moment. There was a previously exisiting TODO around that. 

This PR didn't address that TODO. I think we should let another resolver resolve EC2IMDSRegion instead of having it here as a by product of getRegion. That way if EC2IMDSRegion can be enabled by shared config, all we have to do is implement the interface. Also helps stay consistent. 

We can change this later as all is unexported, so fine either way.  (Same reasoning for  above comment )